pkg/pillar/cmd/zedrouter/probe.go

```diff
@@ -0,0 +1,534 @@
+	} else if info.GatewayUP || info.RemoteHostUP {
+		upCnt = 1
+	}
+	return upCnt
+}
+
+func getRemoteURL(netstatus *types.NetworkInstanceStatus, defaultURL string) string {
+	remoteURL := defaultURL
+	// check on User defined URL/IP address
+	if netstatus.PConfig.ServerURL != "" {
+		if strings.Contains(netstatus.PConfig.ServerURL, "http") {
+			remoteURL = netstatus.PConfig.ServerURL
+		} else {
+			// use 'http' instead of 'https'
+			remoteURL = "http://" + netstatus.PConfig.ServerURL
+		}
+	} else if netstatus.PConfig.ServerIP != nil && !netstatus.PConfig.ServerIP.IsUnspecified() {
+		remoteURL = "http://" + netstatus.PConfig.ServerIP.String()
+	}
+	return remoteURL
+}
+
+func getProbeRatio(netstatus *types.NetworkInstanceStatus) uint32 {
+	if netstatus.PConfig.ProbeInterval != 0 {
+		ratio := netstatus.PConfig.ProbeInterval / NhProbeInteval
+		if ratio < minProbeRatio {
+			return minProbeRatio
+		}
+		return ratio
+	}
+	return remoteTolocalRatio
+}
+
+// base on the probe result, determine if the port should be good to use
+// and record the latency data for reaching the remote host
+func probeProcessReply(info *types.ProbeInfo, gotReply bool, latency int64, isLocal bool) {
+	if isLocal {
+		if gotReply {
+			info.SuccessCnt++
+			info.FailedCnt = 0
+		} else {
+			info.FailedCnt++
+			info.SuccessCnt = 0
+		}
+		if info.FailedCnt > maxContFailCnt && info.GatewayUP {
+			info.GatewayUP = false
+		} else if info.SuccessCnt > maxContSuccessCnt && !info.GatewayUP {
+			info.GatewayUP = true
+		}
+	} else {
+		if gotReply {
+			totalLatency := info.AveLatency * int64(info.SuccessProbeCnt)
+			info.SuccessProbeCnt++
+			info.AveLatency = (totalLatency + latency) / int64(info.SuccessProbeCnt)
+			info.FailedProbeCnt = 0
+		} else {
+			info.FailedProbeCnt++
+			info.SuccessProbeCnt = 0
+			info.AveLatency = 0
+		}
+		if info.FailedProbeCnt > maxContFailCnt && info.RemoteHostUP {
+			info.RemoteHostUP = false
+		} else if info.SuccessProbeCnt > maxContSuccessCnt && !info.RemoteHostUP {
+			info.RemoteHostUP = true
+		}
+	}
+}
+
+func probeFastPing(info types.ProbeInfo) bool {
+	var dstaddress, srcaddress net.IPAddr
+	var pingSuccess bool
+	p := fastping.NewPinger()
+
+	// if we don't have a gateway address or local intf address, no need to ping
+	if info.NhAddr.IsUnspecified() || info.LocalAddr.IsUnspecified() {
+		return false
+	}
+
+	dstaddress.IP = info.NhAddr
+	p.AddIPAddr(&dstaddress)
+
+	srcaddress.IP = info.LocalAddr
+	p.Source(srcaddress.String())
+	if srcaddress.String() == "" || dstaddress.String() == "" {
+		return false
+	}
+	p.MaxRTT = time.Millisecond * time.Duration(maxPingWait)
+	log.Infof("probeFastPing: add to ping, address %s with source %s, maxrtt %v\n",
+		dstaddress.String(), srcaddress.String(), p.MaxRTT)
+	p.OnRecv = func(ip *net.IPAddr, d time.Duration) {
+		if strings.Compare(ip.String(), dstaddress.String()) == 0 {
+			pingSuccess = true
+			log.Infof("probeFastPing: got reply from %s, duration %d nanosec or rtt %v\n",
```

Comment:
This might fill up the logs on device and in cloud. log.Debugf instead?

Reply:
sure. right now is useful for testing, i will remember to remove info into debug when everything is stable.